File: JavaScript/_6_Zigzag_Conversion.js
// The string "PAYPALISHIRING" is written in a zigzag pattern on a given number of rows like this: (you may want to display this pattern in a fixed font for better legibility)

// P   A   H   N
// A P L S I I G
// Y   I   R
// And then read line by line: "PAHNAPLSIIGYIR"

// Write the code that will take a string and make this conversion given a number of rows:

// string convert(string s, int numRows);

// Example 1:

// Input: s = "PAYPALISHIRING", numRows = 3
// Output: "PAHNAPLSIIGYIR"
// Example 2:

// Input: s = "PAYPALISHIRING", numRows = 4
// Output: "PINALSIGYAHRPI"
// Explanation:
// P     I    N
// A   L S  I G
// Y A   H R
// P     I
// Example 3:

// Input: s = "A", numRows = 1
// Output: "A"

// Constraints:

// 1 <= s.length <= 1000
// s consists of English letters (lower-case and upper-case), ',' and '.'.
// 1 <= numRows <= 1000

/**
 * Converts a string into a zigzag pattern across `numRows` rows
 * and then reads it line-by-line to produce a new string.
 *
 * Example for s = "PAYPALISHIRING", numRows = 3:
 * P   A   H   N
 * A P L S I I G
 * Y   I   R
 * Returns: "PAHNAPLSIIGYIR"
 *
 * @param {string} s - The input string to convert.
 * @param {number} numRows - The number of rows to use for the zigzag pattern.
 * @return {string} - The zigzag-converted string read row by row.
 */
var convert = function (s, numRows) {
    // 2D array representing each row of the zigzag
    const zigzag = [];

    // Initialize `numRows` empty arrays (rows) in zigzag
    for (let i = 0; i < numRows; i++) zigzag[i] = [];

    // i: index in input string
    // j: current row index
    // k: current column index (zigzag "vertical" slices)
    let i = 0,
        j = 0,
        k = 0;

    // Process characters until we exhaust the string
    while (s[i]) {
        // Move downward through the rows (vertical part of zigzag)
        while (j < numRows && s[i]) {
            zigzag[j++][k] = s[i++]; // Place char into the current row/col, then go down
        }
        // Step back up diagonally: start one row above bottom
        j = numRows - 2;
        while (j >= 0 && s[i]) {
            zigzag[j--][++k] = s[i++]; // Move up and right in zigzag
            if (j < 0) {               // If we overshoot the top row, reset for next cycle
                j = 1;
                break;
            }
        }
        // Safety check: if diagonal step overshot, fix indices
        if (j < 0) {
            j = 0;
            k++;
        }
    }

    // Read the zigzag rows in order and join into final result
    let result = "";
    for (i = 0; i < zigzag.length; i++) {
        for (j = 0; j < zigzag[i].length; j++) {
            if (zigzag[i][j]) result += zigzag[i][j]; // Skip empty cells
        }
    }

    return result;
};

console.log(convert("AB", 1));

/**
 * Optimized zigzag conversion without creating a full 2D grid.
 * Uses math to jump directly to characters in zigzag order.
 *
 * @param {string} s - The input string to convert.
 * @param {number} numRows - The number of rows to use for the zigzag pattern.
 * @return {string} - The zigzag-converted string read row by row.
 */
var betterConvert = function (s, numRows) {
    // If only one row, zigzag is just the original string
    if (numRows === 1) return s;

    // The full cycle length for one zigzag repetition
    // Example: numRows = 4 → divider = 6
    // Pattern repeats every divider chars
    const divider = 2 + (numRows - 2) * 2;

    let result = "";
    // Outer loop: iterate through each row of zigzag
    for (let i = 0; i <= divider / 2; i++) {
        // Inner loop: jump in steps of `divider` through the string
        for (let j = 0; j < s.length; j += divider) {
            // Vertical column character in this row
            if (j + i < s.length) result += s[j + i];

            // Diagonal character (if applicable):
            // - skip for first and last row
            // - skip if computed index is outside string bounds
            if (i !== 0 && i !== divider - i && j + divider - i < s.length) {
                result += s[j + divider - i];
            }
        }
    }

    return result;
};

console.log(betterConvert("PAYPALISHIRING", 3));

File: Python/_6_Zigzag_Conversion.py
# The string "PAYPALISHIRING" is written in a zigzag pattern on a given number of rows like this: (you may want to display this pattern in a fixed font for better legibility)

# P   A   H   N
# A P L S I I G
# Y   I   R
# And then read line by line: "PAHNAPLSIIGYIR"

# Write the code that will take a string and make this conversion given a number of rows:

# string convert(string s, int numRows);
 

# Example 1:

# Input: s = "PAYPALISHIRING", numRows = 3
# Output: "PAHNAPLSIIGYIR"
# Example 2:

# Input: s = "PAYPALISHIRING", numRows = 4
# Output: "PINALSIGYAHRPI"
# Explanation:
# P     I    N
# A   L S  I G
# Y A   H R
# P     I
# Example 3:

# Input: s = "A", numRows = 1
# Output: "A"
 

# Constraints:

# 1 <= s.length <= 1000
# s consists of English letters (lower-case and upper-case), ',' and '.'.
# 1 <= numRows <= 1000

class Solution:
    def convert(self, s: str, numRows: int) -> str:
        """Converts a string into a zigzag pattern across `numRows` rows and then
        reads it line-by-line to produce a new string.

        Args:
            s (str): the input string to convert
            numRows (int): the number of rows to use for the zigzag pattern

        Returns:
            str: the zigzag-converted string read row by row
        """
        if numRows == 1 or numRows >= len(s):
            return s

        # Create a list for each row
        rows = [''] * numRows
        index, step = 0, 1  # index: current row, step: direction (down or up)

        for char in s:
            rows[index] += char
            # Change direction at the first or last row
            if index == 0:
                step = 1
            elif index == numRows - 1:
                step = -1
            index += step

        # Join all rows to get the final string
        return ''.join(rows)
    
class AnotherSolution:
    def convert(self, s: str, numRows: int) -> str:
        """Converts a string into a zigzag pattern across `numRows` rows and then
        reads it line-by-line to produce a new string.

        Args:
            s (str): the input string to convert
            numRows (int): the number of rows to use for the zigzag pattern

        Returns:
            str: the zigzag-converted string read row by row
        """
        if numRows == 1:
            return s
        
         # Create a list for each row.
        rows = [''] * numRows

        # Initialize variables to track the current row and the direction.
        current_row = 0
        going_down = False

        #Iterate through each character of the input string.
        for char in s:
            # Append the character to the current row.
            rows[current_row] += char

            # Check if we've reached the top or bottom row.
            # If so, reverse the direction.
            if current_row == 0 or current_row == numRows - 1:
                going_down = not going_down

            # Move to the next row on the current direction.
            if going_down:
                current_row += 1
            else:
                current_row -= 1

        # Join the rows to form the final string.
        return "".join(rows)
    
def main():
    sol = Solution()
    sol1 = AnotherSolution()
    print(sol.convert("PAYPALISHIRING", 3))
    print(sol1.convert("PAYPALISHIRING", 4))

if __name__ == "__main__":
    main()